src/utilities/codeBlockHelpers.cy.ts

```diff
@@ -0,0 +1,229 @@
+import { getMatchingLineNumbers, escapeInnerHTML, escapeHTMLIfNeeded, normalizeHighlightedLines, highlightMatchingChars } from './codeBlockHelpers'
+
+const code = `{
+  "compilerOptions": {
+    "target": "es2020",
+    "module": "esnext",
+    "moduleResolution": "node",
+    "allowUnreachableCode": false,
+    "exactOptionalPropertyTypes": true,
+    "noFallthroughCasesInSwitch": true,
+    "noImplicitReturns": true,
+    "noUncheckedIndexedAccess": true,
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
+    "strict": true,
+    "jsx": "preserve"
+  },
+  "include": [
+    "./src",
+    "./types",
+    "./particularly-long-value-that-will-inadvertently-cause-scrolling-for-narrower-containers"
+  ],
+  "markup": "<div class="title">Hello & Hi.</div>",
+}
+`
+
+describe('getMatchingLineNumbers', () => {
+  it('gets matched line numbers by exact match', () => {
+    expect(getMatchingLineNumbers(code, 'true', false)).eql([7, 8, 9, 10, 11, 12, 13])
+    expect(getMatchingLineNumbers(code, ' ', false)).eql([2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21])
+    expect(getMatchingLineNumbers(code, '<', false)).eql([21])
+    expect(getMatchingLineNumbers(code, 'kong', false)).eql([])
+  })
+
+  it('gets matched line numbers by exact match with case sensitivity', () => {
+    expect(getMatchingLineNumbers(code, 'TRUE', false)).eql([])
+  })
+
+  it('gets matched line numbers by regexp match', () => {
+    expect(getMatchingLineNumbers(code, 'tru.', true)).eql([7, 8, 9, 10, 11, 12, 13])
+    expect(getMatchingLineNumbers(code, '[ ]', true)).eql([2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21])
+    expect(getMatchingLineNumbers(code, '.', true)).eql([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22])
+    expect(getMatchingLineNumbers(code, '<[^>]+>', true)).eql([21])
+    expect(getMatchingLineNumbers(code, 'kong', true)).eql([])
+  })
+
+  it('gets matched line numbers by regexp match with case sensitivity', () => {
+    expect(getMatchingLineNumbers(code, 'TRU.', true)).eql([])
+  })
+})
+
+describe('escapeInnerHTML', () => {
+  it('escapes only < and &', () => {
+    expect(escapeInnerHTML('<script>alert("hi")</script>')).eq('&lt;script>alert("hi")&lt;/script>')
+    expect(escapeInnerHTML('<div>&larr; & &rarr;</div>')).eq('&lt;div>&amp;larr; &amp; &amp;rarr;&lt;/div>')
+    expect(escapeInnerHTML('foo')).eq('foo')
+    expect(escapeInnerHTML('')).eq('')
+  })
+})
+
+describe('escapeHTMLIfNeeded', () => {
+  it('escapes only < and &', () => {
+    const regex = /<&>/
+    const escape = cy.spy((v: string) => v)
+
+    expect(escapeHTMLIfNeeded('{ foo: "bar" }', regex, escape)).eq('{ foo: "bar" }')
+    // eslint-disable-next-line @typescript-eslint/no-unused-expressions
+    expect(escape).not.have.been.called
```

Comment:
Just thinking out loud more than anything, but bit annoying there's an API here to make things nicer/easier/shorter to write, but then the linter makes you also add a large comment above it 🤷 😆 

(half noting here as this is related to a completely different piece of work we are doing over in Kuma and it might be handy to link to this at some point, feel free to resolve)

Reply:
I think I'd just leave this as-is for now and may change this if we switch to Vitest at some point.

Reply:
Oh yeah sorry, totally leave as is! It was more of a self-note so I can link to this later for my own purposes.